File: app/presentation/web/csrf.py
import hmac
import secrets

from flask import session

_SESSION_KEY = "csrf_token"


def csrf_token() -> str:
    """Called directly by app/presentation/web/spa.py to embed a fresh token into every served SPA
    shell (as window.__CSRF_TOKEN__), and by validate_csrf() below to check one back."""
    if _SESSION_KEY not in session:
        session[_SESSION_KEY] = secrets.token_urlsafe(32)
    return session[_SESSION_KEY]


def validate_csrf(submitted_token: str | None) -> bool:
    expected = session.get(_SESSION_KEY)
    return expected is not None and submitted_token is not None and hmac.compare_digest(expected, submitted_token)

File: webui/src/api/csrf.ts
declare global {
  interface Window {
    __CSRF_TOKEN__?: string
  }
}

// Injected into the served SPA shell by app/presentation/web/spa.py (serve_spa_shell) — every
// page this app renders (login, change-password, workspace) gets a fresh one on load.
export function csrfToken(): string {
  return window.__CSRF_TOKEN__ ?? ''
}
